### PR TITLE
Add 'makeit' package

### PR DESCRIPTION
**Pull request by arni-magnusson**
Hi, I've just submitted a new package to CRAN called 'makeit', providing a make() function.

Compared to other make-like packages on CRAN, it's the simplest and most straightforward to use for any existing R scripts.

It's just one function that does one thing, but it does it well :) Should fit well as a an import for project management packages with a specific structure and logic.

I took the liberty of proposing an introductory sentence comparing the make-like packages.

Cheers,
Arni

### CODE REVIEW

**Comment by zeileis** [comment]
John @jdblischak did you or one of your co-maintainers have the chance to take a loot at this?

**Comment by wlandau** [comment]
The package uses timestamps to decide whether to rerun a task, so in that sense `makeit` is "Make-like". However, I hesitate to call it a pipeline tool because the scope is limited to a single task. The directed acyclic graph (DAG) is a core concept for pipelines, and it has important practical benefits (resolving execution order, visualizing the graph, and implicit parallel computing). I would say the contribution of `makeit` is much more [`memoise`](https://memoise.r-lib.org/)-like than Make-like.

**Comment by arni-magnusson** [comment]
Thank you for the commentary, much appreciated. The comparison between make-like software is interesting indeed.

In my view, the 'makeit' package offers the same functionality as GNU Make. As such, it provides the building block to construct a pipeline of arbitrary complexity. Similar to writing a Makefile, one can organize make() declarations in a script called _make.R or the like. It's straightforward to transition a simple workflow between a GNU Makefile and 'makeit' declarations, with only minor syntax modifications.

I became curious to find out whether/how the 'makeit' package can handle DAGs, so I put together a couple of worked examples. One is based on the DAG Wikipedia article and the other is based on the DAG example from the 'targets' user manual.

https://github.com/arni-magnusson/makeit/tree/main/inst/examples#examples

At the bottom of this page, I tabulate a comparison between 'makeit' and 'targets'. Feel free to correct any misunderstandings. One key difference seems to be the file-based vs. function-based paradigms to organize a workflow. Overall, though, the two packages are a bit like comparing a brick with a city hall :)

The 'makeit' package seems to fit well in the Pipeline Toolkits section, as it synchronizes computational output with the underlying code and data, and tells the user when everything is up to date. It is similar to GNU Make, but it is R-focused. The package can handle relatively simple pipelines just fine, but other packages are a better choice for medium-to-complex pipelines.

I appreciate the opportunity to study the inner workings of 'targets' package. It's an impressive framework. Also happy that as a spin-off of this study, I now have a useful collection of annotated 'makeit' examples, linked above. I admit the version 1.0.0 help page for make() does not convey clearly the idea that one can organize a collection of make() declarations in a Makefile like format.

**Comment by jdblischak** [comment]
> The 'makeit' package seems to fit well in the Pipeline Toolkits section, as it synchronizes computational output with the underlying code and data, and tells the user when everything is up to date. It is similar to GNU Make, but it is R-focused. The package can handle relatively simple pipelines just fine, but other packages are a better choice for medium-to-complex pipelines.

I agree that makeit belongs in the Pipeline Toolkits section

> I took the liberty of proposing an introductory sentence comparing the make-like packages.

This is the reason I didn't immediately merge the PR. I don't feel like this sentence adds enough new information compared to the existing description in the list below. As you can see in the other sections, we list one very popular package in the introduction as an example of that section (eg only renv is mentioned in the Package Reproducibility section, even though there are many great packages listed in that section). As your makeit package is brand new, I don't think it (yet) warrants discussion in the top-level paragraph. This could of course be revisited in the future if your package becomes widely adopted by the R community

In summary, if you remove the extra sentence and ping me here, I'll merge it right away. Thanks!

**Comment by arni-magnusson** [comment]
> As your makeit package is brand new, I don't think it (yet) warrants discussion in the top-level paragraph.

In its youthful exuberance, the 'makeit' package stepped on a stool in the classroom, pounding its chest with all its 30 lines of code. As its custodian, I feel embarrassed...

I agree 100% and have updated the pull request :)

**Comment by wlandau** [comment]
> The 'makeit' package seems to fit well in the Pipeline Toolkits section, as it synchronizes computational output with the underlying code and data, and tells the user when everything is up to date. 

Maybe for a single step by itself, but not for a collection of steps. For example, if the steps are out of order, then a later task may invalidate an earlier one, and the workflow will never be up to date as a whole. In addition, if two steps are disconnected but are supposed to be connected, there are no guardrails or visualizations to detect this. There is ample room for these and other reproducibility errors if the user is responsible for manually constructing the DAG and manually sorting the steps in topological order.

Pipeline tools protect against these pitfalls by automatically constructing the DAG behind the scenes (using knowledge of all the inputs and outputs of all the steps together). The DAG catches errors, and it ensures that the correct tasks run in the correct order. It also allows users to write tasks in any arbitrary order and trust that the DAG and thus the execution order will still be correct. This is not only a huge convenience, it is also an indispensable piece of the reproducibility claims you cite, and it is why full pipeline tools like Make are allowed to claim a much stronger level of reproducibility than single-step caching/[memoization](https://en.wikipedia.org/wiki/Memoization) tools. With its restriction to a single step, `makeit` is among the latter kind of tool.

`igraph` could quickly and concisely construct a DAG for `makeit` using `igraph::graph_from_data_frame()`. Then you could call `igraph::is_dag()` to make sure the graph is really a DAG. (You may have to remove loops first, but that is straightforward in the data frame upstream of `igraph::graph_from_data_frame()`.) Then, you could easily get the sequential execution order using `igraph::topo_sort()`. Parallel computing on a DAG is possible, but it is an enormous undertaking and probably outside the scope of a deliberately minimalist tool like `makeit`.

**Comment by arni-magnusson** [comment]
Thank you for the insightful comments. The potential complexities from graph theory can be considerable. I reckon there's more to GNU Make than I'm aware of.

For statistical analyses, I sometimes write a short GNU Makefile to rerun time-consuming parts (fitting a model, iterative simulations, etc.) only when necessary. This allows me to update figures and tables quickly after changing some of the data or methods. I've found it straightforward to convert those types of makefiles into R 'makeit' scripts.

One reason for releasing 'makeit' to CRAN was I thought I would probably learn something about pipeline toolkits in general, comparing different approaches and possibly getting feedback from domain experts. Thanks for that!

**Comment by arni-magnusson** [comment]
Ready for merging, I believe

**Comment by jdblischak** [comment]
@arni-magnusson @wlandau Thanks for your discussion above. I took some time to evaluate makeit today. I agree with Will's criticism that makeit does not construct a full DAG, which is a critical feature for ensuring that your pipeline is up-to-date (and thus reproducible). That being said, the package clearly states it is designed to be minimalist, and with the argument `prereq`, a user could constructor a script of calls to `makeit::make()` with very explicit and verbose use of `prereq` to maintain a reproducible pipeline. Specifically, I am envisioning something like the [`dag_wikipedia`](https://github.com/arni-magnusson/makeit/blob/main/inst/examples/dag_wikipedia/_make.R) example where `out/a.dat` is repeated in the `prereq` argument for each call to `make()`.
